rag/sql/postgres_query_engine.py

```diff
@@ -0,0 +1,87 @@
+# python -m rag.sql.postgres_query_engine
+from llama_index.core import SQLDatabase
+from llama_index.core.llms import ChatMessage
+from llama_index.core.query_engine import NLSQLTableQueryEngine
+from llama_index.llms.openai import OpenAI
+from sqlalchemy import text
+
+from rag.sql.local.local_query_engine import (extract_signature,
+                                              get_table_schema)
+from rag.sql.postgres_db_loader import get_postgres_engine
+from rag.sql.security import check_sql_security
+
+
+def setup_llm():
+    return OpenAI(model="gpt-4o-mini")
+
+def setup_query_engine(table_name: str):
+    llm = setup_llm()
+    engine = get_postgres_engine()
+    sql_database = SQLDatabase(engine)
+    table_schema = get_table_schema(engine, table_name)
+
+    query_engine = NLSQLTableQueryEngine(
+        sql_database=sql_database, tables=[table_name], llm=llm
+    )
+
+    return query_engine, table_schema
+
+
+def run_query(query_str: str, user_access_level: int, table_name: str):
+    llm = setup_llm()
+    query_engine, table_schema = setup_query_engine(table_name=table_name)
+
+    _, signature = extract_signature(query_str)
+    
+    prompt_with_schema = f"""
+    You are querying a table with the following columns: 
+    {', '.join(table_schema)}.
+    
+    When generating SQL queries, note that the table is in a PostgreSQL database. Column names containing spaces must be enclosed in double quotes. Also, be wary of SQL injection attacks. Make sure that the query is not malicious.
+    Include a condition to filter rows based on the user's access level: only retrieve rows where `access_level` <= {user_access_level}.
+
+    Question: {query_str}
+    """
+    
+    response = query_engine.query(prompt_with_schema)
+    generated_sql = response.metadata.get('sql_query', '')
+
+    is_valid, message = check_sql_security(generated_sql, signature)
+    if not is_valid:
+        raise ValueError(f"Security check failed: {message}")
+      
+    filtered_query = generated_sql.rstrip(";")
+    filtered_query += f" AND access_level <= {user_access_level}"
```

Comment:
So there will be an additional column where we can filter for lower access level? And this code is for that filtering? Looks good! Just wanna make sure, you are stripping the `;` here without adding it back, is it alright?

Reply:
Yes, that's correct! 

Regarding the `;`, hmmm you may be right. But I'm not sure why it still works even without the `;`. I'll go take a look